### PR TITLE
fix(post-review): degrade unanchorable findings into the review body (#192)

### LINKED ISSUE #192: post_review: findings skipped by line-in-diff validation are silently dropped from PR delivery while the review body still counts them
## Problem

When `post_review.py`'s inline line validation (from the #127/#135 fix) rejects a finding because its anchor line is not in the diff, the finding is skipped with only a stderr warning and a stdout count line — it reaches the PR in no form at all. The review body, composed upstream by the pipeline, still claims the full finding count.

## Live sighting (2026-08-15, PR #191 review)

The pipeline delivered a review whose body said "3 findings survived full verification and blind challenge (1 main, 2 suggestions). See inline comments for details" — but only 2 inline comments were posted. The **main** finding (`cross-file-1`, MEDIUM: README.md's REVIEW.md example not updated to the #94 contract) was anchored at `scripts/filter_findings.py:216`, a file in the diff but a line outside the changed hunks:

```
WARNING: Skipping finding 'README.md's REVIEW.md example was not updated...' at scripts/filter_findings.py:216 — line not found in diff. Valid lines for this file: [70, 71, 72, 73, 74, 75, 76, 150, 151, 152]
Review posted: .../pull/191#pullrequestreview-4944880964
  2 inline comment(s) posted.
  1 finding(s) skipped (lines not in diff).
```

The reviewer reading the PR sees a count that doesn't match the comments, and the highest-severity finding of the run is the one that vanished. It was only recoverable from the on-disk report artifact.

## Contributing factor

The finding's own anchoring was poor — a README defect anchored to a parser file:line (the cross-file-impact agent anchored to the contract site rather than the stale doc site). Better anchoring would have dodged this instance, but delivery must still not silently drop: skip reasons are legitimate and will keep occurring (context lines, files outside the diff, renames).

## Direction

Skipped-by-validation findings should degrade into the review body (or a trailing section of it) instead of disappearing, and the body's stated count should reflect what was actually delivered inline vs appended. Both belong to `post_review.py` so the guarantee is structural — the model composing `review_body` upstream cannot know what validation will later skip.

Fits the Wave 5 delivery cluster (#22/#62/#66/#63 neighborhood) — it changes delivery surface, not discovery/filtering.

## Verification

Per bench/MEASUREMENT.md: suites-only (unit tests over the skip path; parity untouched).

### PR DESCRIPTION
**Pull request by leehopper**
## What

Findings that `post_review.py`'s line-in-diff validation cannot anchor inline no longer vanish from delivery (closes #192, the 2026-08-15 live sighting on PR #191's own review). They now degrade into a trailing **"⚠️ N finding(s) could not be anchored inline"** section on the review body (GitHub) / MR summary note (GitLab), rendered through the same `render_comment_body` path as inline comments, so every finding still reaches the PR/MR.

- **GitHub** (`post_github`): the existing comments-vs-skipped partition feeds `build_skipped_section`, appended to `review_body` before the footer. The section states both counts ("M inline comment(s) were posted; the following N …") — exact, since the single review POST is atomic.
- **GitLab** (`post_gitlab`): the deterministic skips (no line, line not in diff) are pre-partitioned *before* the summary note is composed — both checks are pure functions of already-fetched `valid_lines`, so the decisions and warning strings are byte-identical to what the inline loop previously produced. The note carries the section; the inline loop iterates only the remaining findings with its five-bucket outcome partition intact. The GitLab section deliberately states **no** delivered-inline count: the note posts before the loop, so `already_present`/`invalid`/`failed` outcomes are unknowable at composition time — claiming a number there would be a promise the loop hasn't kept yet. Those two late-loss branches (malformed position, live 400) are out of #192's scope and unchanged.
- Stdout skip line reworded to `N finding(s) skipped inline (lines not in diff) — appended to review body.` — grep confirmed no parser anywhere (bench/, workflows/, skills/, .github/, docs/machine-parsed-strings.md) consumes the old wording.
- Footer semantics untouched: `build_footer(len(findings), …)` still records the total findings count, which stays honest because with this change every finding *is* in the body.

## Hardening (from the pre-PR adversarial review, 2 rounds, 3 findings fixed)

- **Marker forgery (high):** finding `title`/`body` reach the wire raw, and unlike an inline discussion — where the mechanical marker is appended after the prose and always shadows a forgery — nothing mechanical follows a finding's text inside the section. Every `<!--` in the rendered text is neutralized to `&lt;!--`, so a planted `code-gauntlet-finding-key` marker can never parse as a delivery record and suppress a different finding on the next run. Invariant documented in both `post_review.py` and `review_marker.py`.
- **Footer-dedup suppression (medium):** the footer is now composed against the *original* body, then appended after the section — a finding planting footer/marker text for the current sha can no longer talk `build_footer`'s own-signal dedup into omitting the real footer.
- **Overstated GitLab count (medium):** see above — the delivered-count claim is GitHub-only.

## Tests

New: `TestBuildSkippedSection`, `TestGitHubSkippedFindingsDegrade`, `TestGitlabSkippedFindingsDegrade`, `TestSkippedSectionForgeryResistance`, `TestBuildSkippedSectionNoFileNoLine` — section rendering + both-counts, skip lands in the dry-run-captured body and out of the comments payload, footer stays last and mechanical, no-skip body byte-identical, GitLab note carries the section and skipped findings never reach the discussion loop, validation-skipped platform posts everything with no section, redaction reuse, forgery neutralization.

**Mutation-verified red** (with `__pycache__` purged around each): section builder neutered → 5 named tests red; GitHub section-append removed → red; GitLab section-append removed → red; `<!--` neutralization reverted → forgery test red; footer composed against the augmented body → footer test red on its prose-line-count assertion (that assertion exists precisely because the neutralization fix alone masks the marker half of the mutation).

## Verification

Suites-only per the Wave 5 per-PR rule (delivery-surface change; Smoke B observes the wave's accumulated changes): `pytest tests/ -q` → 1314 passed, 513 subtests; coverage gate 93.04% ≥ 91.8; `pytest bench/tests/ -q` → 537 passed; `pre-commit run --all-files` all green. No `workflows/src` change, so no bundle rebuild needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01DkugfQAn4UQTkbf3iu59cs

---

> [!NOTE]
> **Medium Risk**
> Changes PR/MR delivery and marker/idempotency behavior on the review body; hardening is in place but this is security-adjacent delivery surface.
> 
> **Overview**
> **Unanchorable findings are no longer dropped** when line-in-diff validation fails (missing line or line outside the diff). They appear in a trailing **"⚠️ N finding(s) could not be anchored inline"** section on the GitHub review body or GitLab MR summary note, using the same `render_comment_body` path as inline comments.
> 
> On **GitHub**, the existing comments-vs-skipped split feeds `build_skipped_section` before the footer, including exact inline vs skipped counts. On **GitLab**, deterministic skips are **pre-partitioned** before the summary note is built (the note posts first); the inline loop only handles `remaining` findings. GitLab’s section omits a delivered-inline count because that total is not known until after the per-finding loop.
> 
> **Security / idempotency:** `<!--` in the whole skipped section is neutralized so forged finding-key markers cannot parse as delivery records. The mechanical footer is built from the **original** `review_body` and appended after the skipped section so planted footer/marker text cannot suppress the real footer via dedup.
> 
> Docs in `review_marker.py` and module docstrings describe the degraded-section behavior; stdout skip messaging notes findings are appended to the body. New tests cover rendering, GitHub/GitLab integration, forgery resistance, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 770416a731a4d0747470e91315968ac37bd228c4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by leehopper** [comment]
Bugbot's High is fixed in 770416a: neutralization now applies once to the whole composed section at return (`"\n".join(lines).replace("<!--", "&lt;!--")`), closing the finding-controlled `filepath`/`line` heading and any future field by construction. New discriminating test `test_forged_marker_via_filepath_heading_does_not_survive` is mutation-verified: reverting to per-rendering-only neutralization turns it red while the existing body-forgery test stays green.

**Review by leehopper — COMMENTED** [review]
(This review was posted in error during debugging and its content has been removed. See the follow-up review for the actual code-gauntlet findings.)

File: scripts/post_review.py [hunk 1]
```diff
@@ -1021,9 +1095,47 @@ def post_gitlab(data, valid_lines, new_files=None, old_paths=None):
...
+
+        remaining.append((filepath, f))
+
     sha = resolve_marker_sha(data)
     review_body = data.get("review_body", "")
-    review_body += build_footer(len(findings), sha, body=review_body)
+    # Footer computed against the ORIGINAL body — see the matching comment in
+    # post_github for why a skipped finding's raw text must not be able to suppress it.
+    footer = build_footer(len(findings), sha, body=review_body)
+    review_body += build_skipped_section(skipped_entries) + footer
```
Comment (leehopper) [inline scripts/post_review.py:1162]: **💡 [LOW] No adversarial/forgery test for the GitLab skipped-findings path (only GitHub is covered)**

TestSkippedSectionForgeryResistance (tests/test_post_review.py) exercises forged finding-key markers and forged footer/marker text only through the GitHub dry-run path (_DryRunTestBase, platform: github). post_gitlab has its own, differently-timed composition (pre-partition into skipped_entries before the summary note is posted, footer computed against the original review_body at line 1137 before build_skipped_section is appended at line 1138), and GitLab's summary note is posted live, immediately, before the per-finding discussion loop runs — so a forgery bug there is committed to the MR before anything else happens. No test proves the same forged finding-key or forged footer/marker text is neutralized when it goes through post_gitlab's summary-note composition rather than post_github's review-body composition.

**Suggested fix:**
Add a GitLab-flavored variant of the forgery tests using _GitlabLiveRunBase — a finding with a forged `` or forged footer/marker text placed off-diff, asserting the posted summary note (payloads with no 'position' key) neither carries the forged marker in parseable form nor suppresses the real footer/marker.

**Review by leehopper — COMMENTED** [review]
---
Generated by code-gauntlet | Reviewed up to: 770416a731a4d0747470e91315968ac37bd228c4

**Comment by leehopper** [comment]
## 🟡 [MEDIUM] `post_github` crashes (unhandled KeyError) on a finding with no line, contradicting this PR's own docstring

**File:** `scripts/post_review.py` (around lines 917–940 — could not be anchored inline; the line falls outside any hunk in this diff)

Independently flagged by 3 review dimensions (bug-detector, type-design-analyzer, cross-file-impact).

The module docstring, rewritten by this PR, now promises that a finding whose line "cannot be anchored inline (line not in the diff, or no line at all)" degrades gracefully into the trailing skipped section rather than being dropped. `post_gitlab` implements this correctly via `f.get("line")` with an explicit `None` check that routes the finding into `skipped_entries`. `post_github` was never given the equivalent treatment: it still does `line = f["line"]` unconditionally in the per-finding loop, so a finding lacking the `line` key raises an uncaught `KeyError` that aborts the entire run — losing every finding in the batch, not just the one missing a line.

**Suggested fix:** Mirror `post_gitlab`'s guard in `post_github`: use `line = f.get("line")`, and when it is `None`, route the finding into the skipped bucket (with a `warn_skip`) instead of indexing `f["line"]` directly.

**Failure scenario:** A findings JSON with `platform: github` and one finding lacking a `line` key (e.g. a repo-wide finding with only a `file`) causes `post_github` to raise an unhandled `KeyError`, aborting `main()` before ANY comment — including valid inline ones — is posted.

---
_Posted by code-gauntlet (manual fallback — this finding's line could not be anchored inline due to a separate line-anchoring defect in `post_review.py` itself, worth filing as a follow-up issue)._

**Comment by leehopper** [comment]
# code-gauntlet review — full report

## Summary

This PR (issue #192, three commits culminating at `770416a7`) changes how `post_review.py` handles findings that cannot be anchored to an inline comment on GitHub or GitLab. Previously such findings were silently dropped with only a warning printed; now they "degrade" into a new trailing section of the review body (or MR summary note) via a new `build_skipped_section` function, so every finding still reaches the PR/MR even when it can't land inline. Both `post_github` and `post_gitlab` are restructured to partition findings into "remaining" (inline-eligible) and "skipped" (degraded-to-body) sets, with GitLab's partition happening earlier since it posts its summary note before the per-finding loop runs (unlike GitHub's single atomic POST). A substantial portion of the change is defensive: because a finding's title/body/file/line reach the wire unsanitized, the PR neutralizes every `<!--` in the composed skipped section (to prevent forging a delivery marker or mimicking the mechanical footer) and computes the real footer against the original review body before the skipped section is appended. `scripts/review_marker.py` gains documentation-only changes explaining why its "marker is appended last" invariant doesn't hold inside the new skipped-section context. `tests/test_post_review.py` adds a large block of new unit and adversarial tests, including an explicit `TestSkippedSectionForgeryResistance` class.

**Headline result:** 4 high-confidence findings, all of which survived the full validate → filter → challenge pipeline. No findings were degraded to the unverified bucket in this run.

## Main Findings

### 1. [Medium] post_github crashes (unhandled KeyError) on a finding with no line, contradicting this PR's own docstring
- **File:** `scripts/post_review.py:68-940`
- **Description:** The module docstring, rewritten by this PR, now promises that a finding whose line "cannot be anchored inline (line not in the diff, or no line at all)" degrades gracefully into the trailing skipped section rather than being dropped. `post_gitlab` implements this correctly via `f.get("line")` with an explicit `None` check that routes the finding into `skipped_entries`. `post_github` was never given the equivalent treatment: it still does `line = f["line"]` unconditionally in the per-finding loop, so a finding lacking the `line` key raises an uncaught `KeyError` that aborts the entire run — losing every finding in the batch, not just the one missing a line.
- **Suggestion:** Mirror `post_gitlab`'s guard in `post_github`: use `line = f.get("line")`, and when it is `None`, route the finding into the skipped bucket (with a `warn_skip`) instead of indexing `f["line"]` directly.

### 2. [Medium] post_github crashes on a line-less finding the diff's own docstring now promises is handled
- **File:** `scripts/post_review.py:917-940`
- **Description:** The finding record is an untyped dict with no schema validation at the module boundary — `main()` only checks `owner`/`repo`/`pr_number`, never individual finding shape. The two platform consumers now disagree on whether `line` is required: `post_gitlab`'s pre-partition uses `f.get("line")` and degrades a missing line into `build_skipped_section`, matching the diff's updated docstring, while `post_github`'s loop still does `line = f["line"]` (a bare dict subscript, unchanged by this diff), raising an uncaught `KeyError` for the identical case.
- **Suggestion:** Normalize/validate findings into a single structured shape once, before branching by platform (e.g. a small validated record, or at minimum `line = f.get("line")` in `post_github` mirroring `post_gitlab`'s handling), so both platforms honor the same documented contract.
- **Invalid state example:** A findings JSON with `platform: github` and one finding lacking a `line` key (e.g. a repo-wide finding with only a `file`) causes `post_github` to raise an unhandled `KeyError` on line 929, aborting `main()` before any comment — including valid inline ones — is posted.

### 3. [Medium] GitHub path crashes (unhandled KeyError) on a finding with no line — contradicts the diff's own docstring promise, and is untested
- **File:** `scripts/post_review.py:917-940`
- **Description:** The diff broadens the module docstring to promise that a finding whose line "cannot be anchored inline (line not in the diff, or no line at all)" degrades into the trailing section rather than being dropped. `post_gitlab` implements the no-line case via `f.get("line")` (line 1109); `post_github` still does `line = f["line"]` (line 929) with no defensive `.get()` or pre-check, so a GitHub finding lacking a `line` key raises an unhandled `KeyError` inside the per-finding loop, aborting the whole run before ANY finding is posted or captured. No test exercises this input for the GitHub path — only `TestBuildSkippedSectionNoFileNoLine` covers a no-line fixture, and it calls `build_skipped_section` directly rather than going through `post_github`.
- **Suggestion:** Add a GitHub end-to-end test (mirroring `TestGitHubSkippedFindingsDegrade`) with a finding dict that omits `line` entirely and `platform: 'github'`, asserting either graceful degradation into the skipped section or that the failure mode is deliberate and documented — right now neither is verified, and the docstring's blanket claim goes untested for this platform.
- **Failure scenario:** A caller (or a future pipeline change) emits one finding without a resolvable line number for a GitHub-targeted review; instead of that single finding degrading into the review body as the docstring promises, the entire `post_review.py` run throws an unhandled `KeyError` and exits non-zero, silently dropping every finding in the batch, not just the offending one.

*(Findings #1–#3 above are independent corroborations of the same underlying defect from three review dimensions — bug-detector, type-design-analyzer, and cross-file-impact — treat as one fix.)*

### 4. [Low] No adversarial/forgery test for the GitLab skipped-findings path (only GitHub is covered)
- **File:** `scripts/post_review.py:1098-1138`
- **Description:** `TestSkippedSectionForgeryResistance` (in `tests/test_post_review.py`) exercises forged finding-key markers and forged footer/marker text only through the GitHub dry-run path (`_DryRunTestBase`, `platform: github`). `post_gitlab` has its own, differently-timed composition (pre-partitioning findings into `skipped_entries` before the summary note is posted, with the footer computed against the original `review_body` before `build_skipped_section` is appended), and GitLab's summary note is posted live, immediately, before the per-finding discussion loop runs — so a forgery bug there is committed to the MR before anything else happens. No test proves the same forged content is neutralized when it goes through `post_gitlab`'s summary-note composition rather than `post_github`'s review-body composition.
- **Suggestion:** Add a GitLab-flavored variant of the forgery tests using `_GitlabLiveRunBase` — a finding with a forged `<!-- code-gauntlet-finding-key: ... -->` or forged footer/marker text placed off-diff, asserting the posted summary note neither carries the forged marker in parseable form nor suppresses the real footer/marker.
- **Failure scenario:** A subtle ordering bug specific to `post_gitlab` (e.g. computing the footer against `review_body` after the skipped section was appended, unlike the GitHub path) would go undetected because no test exercises forged content through the GitLab code path — the MR's summary note could ship with a suppressed or corrupted delivery marker, breaking prior-review detection on the next rerun.

## Unverified / Pipeline-Degraded

None. All findings from this run cleared the full validate → filter → challenge pipeline and are presented above as high-confidence.

## Review Dimensions Summary

| Dimension | Agent | Findings | Notes |
|-----------|-------|----------|-------|
| Correctness & Error Handling | bug-detector | 1 | 1 medium |
| Security | security-reviewer | 0 | Clean — no findings returned |
| Cross-file Impact | cross-file-impact | 0 | Clean — no findings returned |
| Test Coverage | test-analyzer | 2 | 1 medium, 1 low |
| Conventions & Intent | conventions-and-intent | 0 | Clean — no findings returned |
| Type Design | type-design-analyzer | 1 | 1 medium |
| Code Simplification | code-simplifier | 0 | Clean — no findings returned |

---

**Delivery note:** finding #1's line (940 in `scripts/post_review.py`) fell outside every diff hunk, so `post_review.py`'s own inline-comment posting rejected it with a GitHub 422 ("Line could not be resolved") — an apparent line-anchoring/hunk-resolution defect in the plugin itself, worth a follow-up issue. Finding #4 posted inline at `scripts/post_review.py:1138`.

## Methodology

- Plugin version: 3.11.1 (resolved from plugin cache path)
- Policy: optimized tier (discovery on Sonnet, security-reviewer on Opus), provider: firstParty
- Delivery: tier `all`, no cap
- Review scope: Full (no prior review found for this PR)
- Stats: 11 discovered → 4 high-confidence after validate/filter/challenge, 0 unverified, 0 degraded dimensions

**Comment by leehopper** [comment]
All review findings are addressed in 2ed0511:

- **Findings 1–3 (MEDIUM, `post_github` KeyError on a no-line finding):** the GitHub loop now mirrors GitLab — `f.get("line")`, with `None` degrading into the skipped section (same warning wording), and the section entries are collected at skip time so no bare `f["file"]`/`f["line"]` subscript remains on that path. New end-to-end tests cover a no-line finding (others still post) and a no-line/no-file finding; the boundary-parity test that pinned the old crash is split so `file` (still required with a line) keeps its KeyError pin while `line` now pins graceful degradation.
- **Finding 4 (LOW, GitLab forgery coverage):** new `TestGitlabSkippedSectionForgeryResistance` mirrors the GitHub class through the live GitLab summary-note path — forged marker via body, via the `file` heading, and forged footer/marker not suppressing the real one.
- **Bonus, from this review's own delivery failure:** the 422 that aborted the review POST (a multi-line comment spanning `start_line=68 → line=940`, with 940 outside every hunk) is fixed structurally — `end_line` ranges are now emitted only when every line in `[line, end_line]` is a valid diff line (contiguous ⇒ single hunk, which is what GitHub requires); otherwise the comment falls back to the already-validated single-line anchor instead of losing the whole atomic POST. Behavior with validation skipped is unchanged.

All new mechanisms mutation-verified red; 1324 tests + coverage gate + pre-commit green.
